### PR TITLE
Improve conformance testing

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
This PR aims to improve the conformance test tool by:
- Adding more retry logic to almost everything
  - Adding nodes
  - Cleaning up
  - Test runs
- Add custom tests to cover cloud provider features
  - PVC
  - LB
  - Custom tests will end up in the junit.xml as well
- Replace glog with logrus (interpreting logs with context is far easier - and we have some nice colors now!)
- Add a run script to easily run the command (Uses our vault instance for the required credentials)
- Add Azure & vSphere providers

**Special notes for your reviewer**:
I'm sorry.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/cmd/conformance-tests/scenarios_aws.go [hunk 1]
```diff
@@ -80,6 +80,13 @@ func (s *awsScenario) Cluster(secrets secrets) *v1.Cluster {
 }
 
 func (s *awsScenario) Nodes(num int) []*kubermaticapiv2.Node {
+	os := getOSNameFromSpec(s.nodeOsSpec)
+	ami := ""
+	if os == "coreos" {
+		// TODO: Make machine-controller use latest AMI for CoreOS
+		ami = "ami-0b088568a857b7c27"
+	}
+
```
Comment (mrIncompetent) [inline api/cmd/conformance-tests/scenarios_aws.go]: This must be removed

File: api/cmd/conformance-tests/README.md [hunk 2]
```diff
@@ -2,33 +2,36 @@
...
--cleanup-on-start=true \
--aws-access-key-id=<<AWS_ACCESS_KEY_ID>> \
--aws-secret-access-key=<<AWS_SECRET_ACCESS_KEY_ID>> \
--digitalocean-token=<<DIGITALOCEAN_TOKEN>> \
--hetzner-token=<<HETZNER_TOKEN>> \
--openstack-domain=<<OPENSTACK_DOMAIN>> \
--openstack-tenant=<<OPENSTACK_TENANT>> \
--openstack-username=<<OPENSTACK_USERNAME>> \
--openstack-password=<<OPENSTACK_PASSWORD>>
+./run.sh
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/README.md:34]: @mrIncompetent Can you please also document how to run a subset of the tests?

**Comment by alvaroaleman** [comment]
@mrIncompetent Conflicts again :/

File: api/cmd/conformance-tests/main.go [hunk 3]
```diff
@@ -119,112 +154,228 @@ func main() {
...
+						clusterDeleteLog.Errorf("failed to delete cluster: %v", err)
+					}
+				}(cluster)
 			}
 		}
+		wg.Wait()
 
-		glog.Info("Cleaned up all old clusters")
+		log.Info("Cleaned up all old clusters")
 		os.Exit(0)
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/main.go]: I find it a bit counterintuitive that `cleanup-on-star` means `cleanup and then exit`

File: api/cmd/conformance-tests/main.go [hunk 4]
```diff
@@ -119,112 +154,228 @@ func main() {
...
+	privDER := x509.MarshalPKCS1PrivateKey(privateKey)
+
+	privBlock := pem.Block{
+		Type:    "RSA PRIVATE KEY",
+		Headers: nil,
+		Bytes:   privDER,
+	}
+
+	privatePEM := pem.EncodeToMemory(&privBlock)
+	// Needs to be google_compute_engine as its hardcoded in the kubernetes e2e tests
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/main.go:349]: lol

File: api/cmd/conformance-tests/reports.go [hunk 5]
```diff
@@ -75,19 +66,24 @@ func collectReports(name, reportsDir string, time time.Duration) (*reporters.JUn
...
+	// Only print details errors in case we have a testcase which failed.
+	// Printing everything which has an error will print the errors from retried tests as for each attempt a TestCase entry exists.
+	if report.Failures > 0 || report.Errors > 0 {
+		for _, t := range report.TestCases {
+			status := "PASS"
+			if t.FailureMessage != nil {
+				status = "FAIL"
+			}
+
+			if t.FailureMessage == nil {
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/reports.go:74]: I think you can shorten all of this by doing the continue as the first step in the loop, and omit setting the status variable

File: api/cmd/conformance-tests/runner.go [hunk 6]
```diff
@@ -172,135 +190,328 @@ func (r *testRunner) Run() error {
...
-	r.logInfo(scenario, "Creating cluster: %s...", cluster.Name)
-	cluster, err := r.kubermaticClient.KubermaticV1().Clusters().Create(cluster)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create the cluster: %v", err)
-	}
 	if r.deleteClusterAfterTests {
-		defer r.tryToDeleteCluster(cluster.Name, scenario)
+		defer func() {
+			if err := tryToDeleteClusterWithRetries(log, cluster, r.clusterClientProvider, r.kubermaticClient); err != nil {
+				log.Errorf("failed to delete cluster: %v", err)
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/runner.go:218]: Would be nice to mention _which_ cluster couldn't be deleted :)

File: api/cmd/conformance-tests/runner.go [hunk 7]
```diff
@@ -172,135 +190,328 @@ func (r *testRunner) Run() error {
...
 	apiNodes := scenario.Nodes(r.nodeCount)
-	for i, node := range apiNodes {
-		var keys []*kubermaticv1.UserSSHKey
-		if r.nodeSSHKeyData != nil {
-			keys = append(keys, &kubermaticv1.UserSSHKey{
-				Spec: kubermaticv1.SSHKeySpec{
-					PublicKey: string(r.nodeSSHKeyData),
-				},
-			})
+	if err := r.setupNodes(log, scenario.Name(), cluster, clusterKubeClient, apiNodes, dc); err != nil {
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/runner.go:241]: Naive question: Why is the `nodeCount` a property of the runner? Shouldn'it it be a property of the scenario? I as the person who executes do not care about the node count, I only want it to be correct - And I may not know what `correct` means

File: api/cmd/conformance-tests/runner.go [hunk 8]
```diff
@@ -172,135 +190,328 @@ func (r *testRunner) Run() error {
...
+			return fmt.Errorf("failed to load cloud-config: %v", err)
+		}
+		cmData = cm.Data["config"]
+		return nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to get cloud config ConfigMap: %v", err)
+	}
+
+	filename := fmt.Sprintf("/tmp/%s-cloud-config", cluster.Name)
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/runner.go]: Why is this now below the test home?

File: api/cmd/conformance-tests/runner.go [hunk 9]
```diff
@@ -366,92 +578,181 @@ func (r *testRunner) waitForControlPlane(scenario testScenario, clusterName stri
...
+		}
 
-	scriptBuffer := &bytes.Buffer{}
-	err = tpl.Execute(scriptBuffer, data)
-	if err != nil {
-		return fmt.Errorf("failed to execute test script template: %v", err)
-	}
+		if cluster.Spec.Cloud.AWS != nil {
+			args = append(args, "--provider=aws")
+			args = append(args, fmt.Sprintf("--gce-zone=%s%s", dc.Spec.AWS.Region, dc.Spec.AWS.ZoneCharacter))
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/runner.go:715]: `gce-zone` for aws?

File: api/cmd/conformance-tests/scenarios_aws.go [hunk 10]
```diff
@@ -89,7 +89,7 @@ func (s *awsScenario) Nodes(num int) []*kubermaticapiv2.LegacyNode {
 					AWS: &kubermaticapiv2.AWSNodeSpec{
 						InstanceType: "t2.medium",
 						VolumeType:   "gp2",
-						VolumeSize:   50,
+						VolumeSize:   100,
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/scenarios_aws.go:92]: Why?

File: api/cmd/conformance-tests/scenarios_azure.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+		scenarios = append(scenarios, &azureScenario{
+			version: v,
+			nodeOsSpec: kubermaticapiv2.OperatingSystemSpec{
+				ContainerLinux: &kubermaticapiv2.ContainerLinuxSpec{
+					// Otherwise the nodes restart directly after creation - bad for tests
+					DisableAutoUpdate: true,
+				},
+			},
+		})
+		// CentOS
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/scenarios_azure.go:35]: Why is this disabled everywhere?

File: api/cmd/conformance-tests/scenarios_azure.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	return scenarios
+}
+
+type azureScenario struct {
+	version    *semver.Version
+	nodeOsSpec kubermaticapiv2.OperatingSystemSpec
+}
+
+func (s *azureScenario) Name() string {
```
Comment (alvaroaleman) [inline api/cmd/conformance-tests/scenarios_azure.go:52]: I feel like we could reduce the duplicate code we currently have in the `scenario_$PROVIDER.go` files by a lot

**Review by alvaroaleman — COMMENTED** [review]
Overall looks good :)

I am fine with merging this in a not-perfect state, as its nothing customer facing.

What do you think about adding one provider of your choice in a non-voting manner to CI to get started with that?

File: api/cmd/conformance-tests/main.go (hunk 3 above)
Comment (mrIncompetent) [reply api/cmd/conformance-tests/main.go]: Used that during development, i don't think exiting is necessary anymore.

File: api/cmd/conformance-tests/main.go (hunk 4 above)
Comment (mrIncompetent) [reply api/cmd/conformance-tests/main.go:349]: Recommendation: Don't look into kubernetes e2e code. Just don't.

File: api/cmd/conformance-tests/runner.go (hunk 6 above)
Comment (mrIncompetent) [reply api/cmd/conformance-tests/runner.go:218]: Scenario details (Including the cluster name) exist as context in the log message :)
https://github.com/kubermatic/kubermatic/pull/2258/files#diff-e5bcf5c5b55dd75523db54294cdc1363R204

File: api/cmd/conformance-tests/runner.go (hunk 7 above)
Comment (mrIncompetent) [reply api/cmd/conformance-tests/runner.go:241]: I made the node count configurable via flag. Mainly for the purpose of avoiding potential quota/performance.

I cannot think of any scenario that would require the use a specific node count as any number of nodes should pass the tests.

File: api/cmd/conformance-tests/runner.go (hunk 9 above)
Comment (mrIncompetent) [reply api/cmd/conformance-tests/runner.go:715]: https://github.com/kubermatic/kubermatic/pull/2258#discussion_r228175889

Comment (mrIncompetent) [reply api/cmd/conformance-tests/runner.go:715]: What i'm trying to say: Yes. The AWS code in the e2e tests use that flag.

File: api/cmd/conformance-tests/scenarios_aws.go (hunk 10 above)
Comment (mrIncompetent) [reply api/cmd/conformance-tests/scenarios_aws.go:92]: More speed

File: api/cmd/conformance-tests/scenarios_azure.go (hunk 11 above)
Comment (mrIncompetent) [reply api/cmd/conformance-tests/scenarios_azure.go:35]: They cause errors, which need to be investigated.
I created https://github.com/kubermatic/kubermatic/issues/2303 to cover this

File: api/cmd/conformance-tests/scenarios_azure.go (hunk 12 above)
Comment (mrIncompetent) [reply api/cmd/conformance-tests/scenarios_azure.go:52]: You're right - i'll refactor the files to only include the cloud specs - though in a follow up, if you don't mind

File: api/cmd/conformance-tests/cleanup.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+		if err != nil {
+			return fmt.Errorf("failed to add the annotation '%s=true' to node '%s': %v", eviction.SkipEvictionAnnotationKey, node.Name, err)
+		}
+	}
+
+	machineList, err := clusterAPIClient.ClusterV1alpha1().Machines(metav1.NamespaceSystem).List(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list Machines from cluster: %v", err)
+	}
+	for _, m := range machineList.Items {
```
Comment (kdomanski) [inline api/cmd/conformance-tests/cleanup.go]: Won't deleting the cluster resource first delete the machines within it anyways?

Comment (mrIncompetent) [reply api/cmd/conformance-tests/cleanup.go]: Good point, I changed the order here at some point. 
Since then this is not necessary anymore

**Comment by mrIncompetent** [comment]
@alvaroaleman PTAL

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**